File: src/main/java/me/nasukhov/TukitaLearner/study/StatsRepository.java
package me.nasukhov.TukitaLearner.study;

import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;
import org.springframework.stereotype.Repository;

import java.util.List;

@Repository
public interface StatsRepository extends JpaRepository<Answer, Long> {
    @Query("""
        SELECT new me.nasukhov.TukitaLearner.study.StudentStats(
            a.userId,
            SUM(CASE WHEN a.isCorrect = true THEN 1 ELSE 0 END),
            COUNT(a)
        )
        FROM Answer a
        WHERE a.channelId = :channelId
        GROUP BY a.userId
        ORDER BY COUNT(a) DESC
        """)
    List<StudentStats> getStatsForGroup(@Param("channelId") String channelId);
}

File: src/main/kotlin/me/nasukhov/tukitalearner/study/StatsRepository.kt
package me.nasukhov.tukitalearner.study

import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Query
import org.springframework.data.repository.query.Param
import org.springframework.stereotype.Repository

@Repository
interface StatsRepository : JpaRepository<Answer, Long> {
    @Query(
        """
        SELECT new me.nasukhov.tukitalearner.study.StudentStats(
            a.userId,
            SUM(CASE WHEN a.isCorrect = true THEN 1 ELSE 0 END),
            COUNT(a)
        )
        FROM Answer a
        JOIN GroupQuestion gq
        WHERE gq.group = :groupId
        GROUP BY a.userId
        ORDER BY COUNT(a) DESC
        """,
    )
    fun getStatsForGroup(
        @Param("groupId") groupId: String,
    ): List<StudentStats>
}
